### PR TITLE
add db functions for last checkpoint

### PR DESCRIPTION
**Pull request by jmozah**
This PR adds the DB functions to save and get the last checkpoint along with its slot.

PS: These needs to be used in appropriate place to store the read the checkpoints @potuz

### CODE REVIEW

File: beacon-chain/db/kv/error.go [hunk 1]
```diff
@@ -15,3 +15,6 @@ var ErrNotFoundOriginBlockRoot = errors.Wrap(ErrNotFound, "OriginBlockRoot")
 
 // ErrNotFoundFeeRecipient is a not found error specifically for the fee recipient getter
 var ErrNotFoundFeeRecipient = errors.Wrap(ErrNotFound, "fee recipient")
+
+// errInvalidCheckpoint is returned when the supplied checkoint size is not as expected
+var errInvalidCheckpointSize = errors.New("invalid checkpoint point size")
```
Comment (potuz) [inline beacon-chain/db/kv/error.go:20]:
```suggestion
var errInvalidCheckpointSize = errors.New("invalid checkpoint size")
```

File: beacon-chain/db/kv/error.go [hunk 2]
```diff
@@ -15,3 +15,6 @@ var ErrNotFoundOriginBlockRoot = errors.Wrap(ErrNotFound, "OriginBlockRoot")
 
 // ErrNotFoundFeeRecipient is a not found error specifically for the fee recipient getter
 var ErrNotFoundFeeRecipient = errors.Wrap(ErrNotFound, "fee recipient")
+
+// errInvalidCheckpoint is returned when the supplied checkoint size is not as expected
```
Comment (potuz) [inline beacon-chain/db/kv/error.go:19]:
```suggestion
// errInvalidCheckpointSize is returned when the supplied checkpoint size is invalid
```

File: beacon-chain/db/kv/validated_checkpoint.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+func (s *Store) LastValidatedCheckpoint(ctx context.Context) ([32]byte, types.Slot, error) {
+	ctx, span := trace.StartSpan(ctx, "BeaconDB.LastValidatedCheckpoint")
+	defer span.End()
+
+	var lastChkPoint [32]byte
+	var slot types.Slot
+	err := s.db.View(func(tx *bolt.Tx) error {
+		bkt := tx.Bucket(lastValidatedCheckpoint)
+		val := bkt.Get([]byte("lastChkPoint"))
+		if len(val) != 40 { // 32 for checkpoint root + 8 bytes for slot
```
Comment (potuz) [inline beacon-chain/db/kv/validated_checkpoint.go:22]: Better make this number 40 a named constant, something like `validatedCheckpointSize`

**Review by terencechain — CHANGES_REQUESTED** [review]
Checkpoint typically refers to the following construct:
```
class Checkpoint(Container):
    epoch: Epoch
    root: Root
```

It's confusing to call a `([32]byte, types.Slot)` pair as a checkpoint. However, I also don't think the new fork choice requires `slot` as an input, and the caller can get the slot from the state summary anyway. So I would highly recommend we simplify the methods down to:

```
LastValidatedRoot(ctx context.Context) ([32]byte, error)
SaveLastValidatedRoot(ctx context.Context, checkPoint [32]byte) error
```

File: beacon-chain/db/kv/validated_checkpoint.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+			return errInvalidCheckpointSize
+		}
+		lastChkPoint = bytesutil.ToBytes32(val[:32])
+		slot = types.Slot(binary.LittleEndian.Uint64(val[32:]))
+		return nil
+	})
+	return lastChkPoint, slot, err
+}
+
+func (s *Store) SaveLastValidatedCheckpoint(ctx context.Context, checkPoint [32]byte, slot types.Slot) error {
```
Comment (potuz) [inline beacon-chain/db/kv/validated_checkpoint.go:32]: Given that Checkpoint means something in the spec (typically a pair of an epoch and a root), can we call this parameter `root`?

File: beacon-chain/db/kv/validated_checkpoint.go [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+	ctx, span := trace.StartSpan(ctx, "BeaconDB.saveLastValidatedCheckpoint")
+	defer span.End()
+
+	updateErr := s.db.Update(func(tx *bolt.Tx) error {
+		value := make([]byte, 40)
```
Comment (potuz) [inline beacon-chain/db/kv/validated_checkpoint.go:37]: Same here, use the named constant

File: beacon-chain/db/kv/validated_checkpoint_test.go [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+import (
+	"bytes"
+	"context"
+	"testing"
+
+	types "github.com/prysmaticlabs/eth2-types"
+	"github.com/prysmaticlabs/prysm/testing/require"
+)
+
+func TestValidateCheckpoint(t *testing.T) {
```
Comment (potuz) [inline beacon-chain/db/kv/validated_checkpoint_test.go:12]:
```suggestion
func TestValidatedCheckpoint(t *testing.T) {
```

File: beacon-chain/db/kv/validated_checkpoint_test.go [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+	slotB := types.Slot(2)
+
+	// add first checkpoint
+	require.NoError(t, db.SaveLastValidatedCheckpoint(ctx, checkpointA, slotA))
+
+	rcvdRoot, rcvdSlot, err := db.LastValidatedCheckpoint(ctx)
+	require.NoError(t, err)
+
+	require.Equal(t, 0, bytes.Compare(checkpointA[:], rcvdRoot[:]))
+	require.Equal(t, true, uint64(slotA) == uint64(rcvdSlot))
```
Comment (potuz) [inline beacon-chain/db/kv/validated_checkpoint_test.go:29]: Why not `require.Equal(t, slotA, rcvdSlot)`?

File: beacon-chain/db/kv/validated_checkpoint.go [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+	ctx, span := trace.StartSpan(ctx, "BeaconDB.LastValidatedCheckpoint")
+	defer span.End()
+
+	var lastChkPoint [32]byte
+	var slot types.Slot
+	err := s.db.View(func(tx *bolt.Tx) error {
+		bkt := tx.Bucket(lastValidatedCheckpoint)
+		val := bkt.Get([]byte("lastChkPoint"))
+		if len(val) != 40 { // 32 for checkpoint root + 8 bytes for slot
+			return errInvalidCheckpointSize
```
Comment (potuz) [inline beacon-chain/db/kv/validated_checkpoint.go:23]: I think we should default to the last finalized checkpoint if there was no checkpoint in DB, the first time we do the migration this should happen, perhaps log a warning and set the db with that value in this function?

**Review by potuz — CHANGES_REQUESTED** [review]

File: beacon-chain/db/kv/validated_checkpoint.go (hunk 8 above)
Comment (jmozah) [reply beacon-chain/db/kv/validated_checkpoint.go:23]: As @terencechain said above, `FinalizedCheckpoint` returns a `root` and an `Epoch`. How do i get slot ?

**Pull request closed without merging**